File: src/main/java/tj/mirboboev/sort/InsertionSort.java
package tj.mirboboev.sort;

import tj.mirboboev.util.Out;

class InsertionSort {

    public static void main(String[] args) {
        Out.array(sort(new Integer[]{6, 7, 8, 1, 4}));
        Out.array(sort(new Integer[]{}));
        Out.array(sort(new Integer[]{6}));
    }

    private static Integer[] sort(Integer[] unsorted) {
        for (int i = 1; i < unsorted.length; i++) {
            int j = i;

            while (j > 0 && (unsorted[j] < unsorted[j - 1])) {
                swap(unsorted, j, j - 1);

                j = j - 1;
            }
        }

        return unsorted;
    }

    private static <T> void swap(T[] array, int i, int j) {
        T temp = array[j];
        array[j] = array[i];
        array[i] = temp;
    }
}

File: src/main/kotlin/sort/InsertionSort.kt
package sort

import tj.mirboboev.util.Out

class InsertionSort {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            Out.array(sort(arrayOf(1, 66, 55)))
            Out.array(sort(arrayOf()))
            Out.array(sort(arrayOf(100)))
        }

        private fun sort(set: Array<Int>): Array<Int> {
            var i = 1

            while (i < set.size) {
                var j = i

                while (j > 0 && (set[j] < set[j - 1])) {
                    set[j] = set[j - 1].also { set[j - 1] = set[j] }
                    j -= 1
                }

                i++
            }

            return set;
        }
    }
}
